components/ai_chat/content/browser/ai_chat_brave_search_throttle.h

```diff
@@ -0,0 +1,71 @@
+/* Copyright (c) 2024 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#ifndef BRAVE_COMPONENTS_AI_CHAT_CONTENT_BROWSER_AI_CHAT_BRAVE_SEARCH_THROTTLE_H_
+#define BRAVE_COMPONENTS_AI_CHAT_CONTENT_BROWSER_AI_CHAT_BRAVE_SEARCH_THROTTLE_H_
+
+#include <memory>
+#include <string>
+
+#include "base/memory/raw_ptr.h"
+#include "base/memory/weak_ptr.h"
+#include "brave/components/ai_chat/core/browser/ai_chat_service.h"
+#include "content/public/browser/navigation_throttle.h"
+#include "content/public/browser/permission_result.h"
+
+namespace content {
+class WebContents;
+}
+
+class PrefService;
+
+namespace ai_chat {
+
+// A network throttle which intercepts Brave Search requests.
+// Currently the only use case is to intercept requests to open Leo AI chat, so
+// it is only created when navigating to open Leo button URL from Brave Search.
+// It could be extended to other Brave Search URLs in the future.
+//
+// For Open Leo feature, we check:
+// 1) If AI chat is enabled.
+// 2) If the request is from Brave Search and is trying to navigate to open Leo
+// button URL.
+// 3) If the nonce property in the a tag element is equal to the one in url ref.
+// 4) If the user has granted permission to open Leo.
+// The navigation to the specific Open Leo URL will be cancelled, and Leo AI
+// chat will be opened only if all the above conditions are met.
+class AIChatBraveSearchThrottle : public content::NavigationThrottle {
+ public:
+  AIChatBraveSearchThrottle(
+      base::OnceCallback<void(content::WebContents*)> open_leo_delegate,
+      content::NavigationHandle* handle,
+      AIChatService* ai_chat_service);
+  ~AIChatBraveSearchThrottle() override;
+
+  static std::unique_ptr<AIChatBraveSearchThrottle> MaybeCreateThrottleFor(
+      base::OnceCallback<void(content::WebContents*)> open_leo_delegate,
+      content::NavigationHandle* navigation_handle,
+      AIChatService* ai_chat_service,
+      PrefService* pref_service);
+
+  ThrottleCheckResult WillStartRequest() override;
+  const char* GetNameForLogging() override;
+
+ private:
+  void OnGetOpenLeoButtonNonce(const std::optional<std::string>& nonce);
+  void OnPermissionPromptResult(blink::mojom::PermissionStatus status);
+  void OnOpenLeo();
+
+  void OpenLeoWithStagedConversations();
+
+  base::OnceCallback<void(content::WebContents*)> open_leo_delegate_;
```

Comment:
`show_conversation_for_content_delegate_`?

Reply:
Renamed to open_ai_chat_delegate_ in https://github.com/brave/brave-core/commit/ab4ea07f080379399d945e2357de36394ecf6fc0.